cmd/uhc/cluster/login/login.go

```diff
@@ -152,3 +168,69 @@ func run(cmd *cobra.Command, argv []string) {
 		os.Exit(1)
 	}
 }
+
+// doSurvey will ask user to choose one if there are more than one clusters match the query
+func doSurvey(clusters []*v1.Cluster) (cluster *v1.Cluster, err error) {
+	clusterList := []string{}
+	for _, v := range clusters {
+		clusterList = append(clusterList, fmt.Sprintf("Name: %s, ID: %s", v.Name(), v.ID()))
+	}
+	choice := ""
+	prompt := &survey.Select{
+		Message: "Please choose a cluster:",
+		Options: clusterList,
+		Default: clusterList[0],
+	}
+	err = survey.AskOne(prompt, &choice, func(ans interface{}) error {
+		choice := ans.(string)
+		found := false
+		for _, v := range clusters {
+			if strings.Contains(choice, v.ID()) {
+				found = true
+				cluster = v
+			}
+		}
+		if !found {
+			return fmt.Errorf("the cluster you choose is not valid: %s", choice)
+		}
+		return nil
+	})
+	return cluster, err
+}
+
+// findClusters finds the clusters that match the given key. A cluster matches the key if its
+// identifier is that key, or if its name starts with that key. For example, the key `prd-2305`
+// doesn't match a cluster directly because it isn't a valid identifier, but it matches all clusters
+// whose names start with `prd-2305`.
+func findClusters(collection *v1.ClustersClient, key string) (clusters []*v1.Cluster, err error) {
+
+	// Get the resource that manages the cluster that we want to display:
+	clusterResource := collection.Cluster(key)
+	response, err := clusterResource.Get().Send()
+
+	// If it's not an cluster id, try to query clusters using search param, we assume the
+	// list should be within 50, if not, we need update the cluster names prefix to minimize
+	// the list
+	if err != nil {
+		pageSize := 50
+		pageIndex := 1
+		listRequest := collection.List().
+			Size(pageSize).
+			Page(pageIndex)
+		listRequest.Search("name like '" + key + "'")
+		listResponse, err := listRequest.Send()
+		if err != nil {
+			fmt.Fprintf(os.Stderr, "Can't retrieve clusters: %s\n", err)
+			return clusters, err
+		}
+		listResponse.Items().Each(func(cluster *v1.Cluster) bool {
+			clusters = append(clusters, cluster)
+			return true
+		})
+		return clusters, nil
+	}
+	cluster := response.Body()
+	clusters = append(clusters, cluster)
+	return clusters, nil
```

Comment:
It is a bit weird to have this code here. When I first read it I thought that it will always be executed, but then I realized that there is a `return` inside the previous block. I'd suggest to move this up, and make an earlier return when you already found the cluster by name:

```go
response, err := clusterResource.Get().Send()

// If there was no error then we already found the cluster and
// we can just return it to the user:
if err == nil {
        cluster := response.Body()
        clusters = []*v1.Cluster{cluster}
        return
}
```